### ISSUE TITLE
Saving a pubspec.yaml file results in re-computed analysis contexts

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#51173]
I actually don't have personal evidence of this, but the team seemed to confirm it is true. It also sounded like we don't know why analysis contexts _should_ be re-computed when saving a `pubspec.yaml` file.

I'm also assuming that re-computing analysis contexts is (and should be) unrelated to invalidating cached diagnostics for each library.

CC @scheglov @bwilkerson @jacob314

**Comment by bwilkerson** [comment]
Yes. I believe that there is no reason to recompute analysis contexts when a `pubspec.yaml` file is edited.

I'd also like to propose that we only invalid the Dart files in a context when a `pubspec.yaml` file is done being edited (that is, only on save, never on a keypress) __and__ the lint in question has been enabled.

The latter part sounds pretty special-cased, but we could generalize it somewhat by having a getter on `LintRule` that returns true if the results of the lint depend on the content of the `pubspec.yaml` file. At the moment there's only one rule with that characteristic, but it would allow for the addition of others in the future.

**Comment by incendial** [comment]
Hm, I actually though you do it on purpose to invalidate possible element model changes, if a dependency was updated. You do it some other way?

**Comment by srawlins** [comment]
> After `dart pub get / upgrade`?

Yes, this is the way.
